book/src/internals/type-system.md

```diff
@@ -0,0 +1,224 @@
+
+An SQL row would translate to struct, and a relation would translate to a list
+of structs.
+
+I would also strive for the type system to be minimal - don't differentiate
+between tuples, objects and structs. Choose one and stick to it.
+
+**Type constraints** - constrain a type with a predicate. For example, have a
+type of int64 that are equal or greater than 10. Postgres
+[does support this](https://news.ycombinator.com/item?id=34835063). The primary
+value of using constrained types would not be validation (as it is used in
+linked article), but when matching the type.
+
+Say, for example, that we have a pipeline like this:
+
+```
+derive color = switch [x => 'red', true => 'green']
+derive is_red = switch [color == 'red' => true, color == 'green' => false]
+```
+
+It should be possible to infer that `color` is of type `text`, but only when
+equal to `'red'` or `'green'`. This means that the second switch covers all
+possible cases and `is_red` cannot be `null`.
+
+## Physical layout
+
+_Logical type_ is user-facing the notion of a type that is the building block of
+the type system.
+
+_Physical layout_ is the underlying memory layout of the data represented by a
+variable.
+
+In many programming languages, physical layout of a logical type is dependent on
+the target platform. Similarly, physical layout of a PRQL logical type is
+dependent on representation of that type in target STS.
+
+```
+PTS logical type  --->  STS logical type  ---> STS physical layout
+```
+
+Note that STS types do not have a single physical layout. Postgres has a logical
+type `anyelement`, which is a super type of any data type. It can be used a
+function parameter type, but does not have a single physical layout so it cannot
+be used a in column declarations.
+
+For now, PRQL does not define a physical layout of any type. It is not needed
+since PRQL is not used for DDL (see section "Built-in primitives") or does not
+support raw access to underlying memory.
+
+As a consequence, results of a PRQL query cannot be robustly compared across
+DBMSs, since the physical layout of the result will vary.
+
+In the future, PRQL may define a common physical layout of types, probably using
+Apache Arrow.
+
+## Syntax
+
+```
+# built-in types
+type int
+type float
+type bool
+type text
+type char
+type null
+
+# users-defined types
+type my_int = int
+
+# sum type (union or enum)
+type number = int | float
+type scalar = number | bool | str | char
+
+# by default types are not nullable
+type my_nullable_int = int | null
+
+# user-defined enum
+type open
+type pending
+type closed
+type status = open | pending | closed
+
+# product type
+type my_struct = [id = int_nullable, name = str]
+type my_timestamp_with_timezone = [int, text]
+
+# list type (I'm not sure about this syntax)
+type list = {T}
+
+# relations are list of structs
+type my_relation = {[
+	id = int,
+	title = text,
+	age = int
+]}
+
+type invoices = {[
+    invoice_id = int64,
+    issued_at = timestamp,
```

Comment:
> @vanillajonathan 
> datetime instead of timestamp?
>
> As for timestamp maybe this should be named datetime?
> It is datetime in Microsoft SQL Server ([documentation](https://learn.microsoft.com/en-us/sql/t-sql/data-types/datetime-transact-sql?view=sql-server-ver16)) which also offers the date and timedata types, so if we were to offer a date and time data type (which I think we should) then datetime would be much more suitable, coherent and predictable than timestamp.

I'm torn on this question.

datetime is more user-friendly, but also a bit misleading in that it refers to a date and a time including a timezone, when in fact it is refering to a moment, which is timezone independent.

My feeling here is that we should find a good (modern) API that deals with time and build on top of their findings.

Reply:
A solid API that hasn't disappointed me yet is [chrono crate](https://docs.rs/chrono/0.4.23/chrono/index.html)

Also Java 8 [java.time package](https://docs.oracle.com/javase/8/docs/api/java/time/package-summary.html)

Reply:
Fully agreed on using a good modern API for datetime logic. A major source of logic bugs in data transformations comes from datetime handling.

Chrono is definitely a great reference and personally I think there should definitely be at least two native datetime types for sql usage: datetime (with TZ) and naive_datetime (without TZ).

Reply:
Filed here: https://github.com/PRQL/prql/issues/2125